### DISCUSSION TITLE
Set note on time

### DISCUSSION THREAD

**Discussion by drovak** [mido/mido#403] [General]
I have a MIDI glockenspiel. Because of the way the solenoid drivers are constructed, a note needs to be off for a certain amount of time before it can sound again. Hence, I'd like to import MIDI files and set all note lengths to the same, short value (since the glockenspiel cannot sustain notes).

My idea was to import a MIDI file, convert all message timestamps from relative to absolute, remove all note_off messages, generate all new note_off messages with an absolute timestamp some delta from its corresponding note_on message, sort all messages by absolute timestamp, and recompute the relative timestamps. 

Is there a trivial way to accomplish this?

**Comment by rdoursenaud** [comment]
Yes!
I would combine samples from the documentation:
   - [Creating a New File](https://mido.readthedocs.io/en/latest/midi_files.html#creating-a-new-file)
   - [Iterating Over Messages](https://mido.readthedocs.io/en/latest/midi_files.html#iterating-over-messages)

The result would look something like this (Untested code written on the fly):
```python
from mido import Message, MidiFile, MidiTrack

# Adjust to taste
NOTE_OFF_DELAY = .1  # 100 ms

# Let's prepare our target file
mid = MidiFile()
track = MidiTrack()
mid.tracks.append(track)

# Parse the original file
for msg in MidiFile('orig.mid'):
    if msg.type == 'note_on':
        if msg.velocity !=0:  # 0 velocity messages also means Note Off
            # We keep the Note On message
            track.append(msg)
            # Let's build a Note Off just after the Note On
            note_off = Message('note_off', note=msg.note, velocity=0, time=NOTE_OFF_DELAY)
            track.append(note_off)
        else:
            pass  # We omit the zero velocity Note On since they're Note Off and we build our own.
    elif msg.type != 'note_off':  # Let's keep all the rest that is neither Note On nor Note Off
        track.append(msg)

# Finally save our modified file
mid.save('mod.mid')



```

**Reply by drovak** [reply]
Thanks! This is starting to make some sense, but I can't make heads and tails of the time value just yet. Running your program seems to be generating non-integer values for `msg.time`.

**Reply by rdoursenaud** [reply]
I made a mistake.
With `mido`, `time` defines the delta-time in seconds from the previous message.
I edited my answer above to reflect this.

Changes:

    NOTE_OFF_DELAY value from 1 to .1
    Set the note_off time to NOTE_OFF_DELAY instead instead of msg.time + NOTE_OFF_DELAY

**Reply by drovak** [reply]
This is still not working. I've boiled it down to this simple example, which I expect would reproduce the input file exactly. This still doesn't work.

```
mid = MidiFile()
track = MidiTrack()
mid.tracks.append(track)
orig = MidiFile('old.mid')

for msg in orig:
    track.append(msg)

mid.save('new.mid')
```

It seems like when messages are read from the file, the time value is in seconds, but when appending a message, the time value is expected to be in ticks. This is not intuitive.

**Reply by rdoursenaud** [reply]
Hum, that's what I get for not testing my code.
I had something like that in the back of my head but didn't remember it was *that* confusing.
Yes, the `time` parameter is ill defined and changes unit and/or meaning depending on usage and/or origin.
This is a design mistake that has not been addressed yet.
